multidict/_multidict_c.c

```diff
@@ -0,0 +1,854 @@
+{
+    Py_ssize_t pos1 = 0,
+               pos2 = 0;
+
+    Py_hash_t h1 = 0,
+              h2 = 0;
+
+    PyObject *identity1 = NULL,
+             *identity2 = NULL,
+             *value1    = NULL,
+             *value2    = NULL;
+
+    if (self == other) {
+        Py_RETURN_TRUE;
+    }
+
+    if (pair_list_len(self->impl) != pair_list_len(other->impl)) {
+        Py_RETURN_FALSE;
+    }
+
+    while (_pair_list_next(self->impl, &pos1, &identity1, NULL, &value1, &h1) &&
+           _pair_list_next(other->impl, &pos2, &identity2, NULL, &value2, &h2))
+    {
+        if ((h1 != h2) ||
+            PyObject_RichCompare(identity1, identity2, Py_NE) ||
+            PyObject_RichCompare(value1, value2, Py_NE))
+        {
+            Py_RETURN_FALSE;
+        }
+    }
+
+    Py_RETURN_TRUE;
+}
+
+static INLINE int
+_multidict_update_items(_MultiDictObject *self, _MultiDictObject *impl)
+{
+    return pair_list_update((PyObject*)self->impl, (PyObject*)impl);
+}
+
+static int
+_multidict_append_items(_MultiDictObject *self, _MultiDictObject *impl)
+{
+    PyObject *key   = NULL,
+             *value = NULL;
+
+    Py_ssize_t pos = 0;
+
+    while (_pair_list_next((PyObject*)impl, &pos, NULL, &key, &value, NULL)) {
+        if (pair_list_add(self->impl, key, value) < 0) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int
+_multidict_append_items_seq(_MultiDictObject *self, PyObject *arg,
+                            const char *name)
+{
+    PyObject *key   = NULL,
+             *value = NULL,
+             *item  = NULL,
+             *iter  = PyObject_GetIter(arg);
+
+    if (iter == NULL) {
+        return -1;
+    }
+
+    while ((item = PyIter_Next(iter)) != NULL) {
+        if (PyObject_Length(item) != 2) {
+            PyErr_Format(
+                PyExc_TypeError,
+                "%s takes either dict or list of (key, value) tuples",
+                name,
+                NULL
+            );
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+
+        key   = PyTuple_GET_ITEM(item, 0);
+        value = PyTuple_GET_ITEM(item, 0);
+        pair_list_add(self->impl, key, value);
+
+        Py_DECREF(item);
+    }
+
+    Py_DECREF(iter);
+    return 0;
+}
+
+static int
+_multidict_list_extend(PyObject *list, PyObject *target_list)
+{
+    PyObject *item = NULL,
+             *iter = PyObject_GetIter(target_list);
+
+    if (iter == NULL) {
+        return -1;
+    }
+
+    while ((item = PyIter_Next(iter)) != NULL) {
+        if (PyList_Append(list, item) < 0) {
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+        Py_DECREF(item);
+    }
+
+    Py_DECREF(iter);
+    return 0;
+}
+
+static int
+_multidict_extend(_MultiDictObject *self, PyObject *args, PyObject *kwds,
+                  const char *name, int do_add)
+{
+    // TOOD: Refactoring me. Split on little functions;
```

Comment:
The function needs a love, I see many TODOs and agree with them

Reply:
Partially done. Please look new implementation.